collect_app/src/main/java/org/odk/collect/android/utilities/PermissionUtils.java

```diff
@@ -88,13 +88,55 @@ public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissio
                 .check();
     }
 
+    public static void requestLocationPermissions(@NonNull Activity activity, @NonNull PermissionListener action) {
+        MultiplePermissionsListener multiplePermissionsListener = new MultiplePermissionsListener() {
+            @Override
+            public void onPermissionsChecked(MultiplePermissionsReport report) {
+                if (report.areAllPermissionsGranted()) {
+                    action.granted();
+                } else {
+
+                    AlertDialog.Builder builder = new AlertDialog.Builder(activity, R.style.Theme_AppCompat_Light_Dialog);
```

Comment:
Just a suggestion:
It will be good just to call `action.denied` and show the alert at the call site.
By showing the alert here, `SRP` is violated as the `PermissionUtils` is doing 2 things:
Checking for permissions and showing alert.
Also it would be easier to test if the dialog is shown or not by moving out the code to display the alert.

Reply:
I agree `PermissionUtils` doesn't have an ideal structure -- it's a grab bag of static methods that do a lot of things (this is really `requestLocationPermissionsIfNotGrantedAndShowDialogIfDenied`). I don't have a great alternate suggestion that still reduces redundancy a bit between the two places this is used. @vishnus1224 do you have a specific approach you'd recommend? You can always try something out in a branch and share a link to it or put something in a gist as a demo.

Luckily testing is not an issue, see https://github.com/opendatakit/collect/pull/2320#discussion_r199594982

Reply:
> It will be good just to call action.denied and show the alert at the call site.

Actually, this is a pretty good approach. Ideally this static method wouldn't even be needed. But we can leave it this way for now and move it once the redundancy between the geo classes is addressed.

Reply:
Thanks for the points @vishnus1224 I agree @lognaturel with the whole notion that there are tons of static methods doing a lot. At the time of implementation, I couldn't come up with a simpler way to make the code as reusable as it is now so that's the reason I opted for this approach. Also, most permissions code is boilerplate and doesn't change that much so I just thought putting it all in one place so it can easily be accessed from all the necessary places would make it easier to implement and manage. Nonetheless, I am definitely gonna take a look at it again!

Reply:
Ok, so perhaps we can test this one, get it ready for merging and then if there are good ideas for a restructure, do it in a separate PR.

Reply:
Yes, that sounds like a good idea.